src/Interactive/Features/Interactive/Core/InteractiveHost.cs

```diff
@@ -231,21 +232,24 @@ private bool CheckAlive(Process process)
 
         public void Dispose()
         {
-            Dispose(joinThreads: false, disposing: true);
+            Dispose(joinThreads: true, disposing: true);
         }
 
         private void Dispose(bool joinThreads, bool disposing)
         {
-            if (disposing)
+            using (_disposeLock.DisposableWait())
             {
```

Comment:
What's the benefit of using DisposableWait, vs. just ```_disposeLock.Wait(); _disposeLock.Release();```?

Reply:
Also, we should probably not wait when running in finalizer (disposing == false).

---
In reply to: [160765822](https://github.com/dotnet/roslyn/pull/24058#discussion_r160765822) [](ancestors = 160765822)

Reply:
One more thing: Since we access _disposeLock in destructor we should guard against it being null, in case the constructor of this class doesn't run to completion.

```
var semaphore = _disposeSemaphore;
if (semaphore != null)
{
  semaphore.Wait();
  semaphore.Dispose();
}
```


---
In reply to: [160766053](https://github.com/dotnet/roslyn/pull/24058#discussion_r160766053) [](ancestors = 160766053,160765822)